polygeist/lib/Dialect/Polygeist/Transforms/FunctionSpecialization.cpp

```diff
@@ -0,0 +1,231 @@
+  using FunctionSpecializationBase<
+      FunctionSpecializationPass>::FunctionSpecializationBase;
+
+  void runOnOperation() override;
+
+private:
+  /// Returns true if \p func is a candidate.
+  bool isCandidateFunction(FunctionOpInterface func) const;
+  /// Returns true if \p acc1 and \p acc2 need to be checked for no overlap.
+  bool isCandidateAccessorPair(AccessorType acc1, AccessorType acc2) const;
+  /// Populate \p accessorPairs with accessor pairs that should be checked for
+  /// no overlap for \p call.
+  void getRequireNoOverlapAccessorPairs(
+      CallOpInterface call,
+      SmallVectorImpl<AccessorPairType> &accessorPairs) const;
+  /// Version \p call.
+  void versionCall(CallOpInterface call) const;
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// FunctionSpecializationPass
+//===----------------------------------------------------------------------===//
+
+bool FunctionSpecializationPass::isCandidateAccessorPair(
+    AccessorType acc1, AccessorType acc2) const {
+  if (acc1 == acc2)
+    return false;
+  if (relaxedAliasing)
+    return true;
+  auto acc1Ty = cast<sycl::AccessorType>(acc1.getType().getElementType());
+  auto acc2Ty = cast<sycl::AccessorType>(acc2.getType().getElementType());
+  return (acc1Ty.getType() == acc2Ty.getType());
+}
+
+void FunctionSpecializationPass::getRequireNoOverlapAccessorPairs(
+    CallOpInterface call,
+    SmallVectorImpl<AccessorPairType> &accessorPairs) const {
+  SmallVector<AccessorType> candArgs;
+  collectCandidateArguments(call, candArgs);
+  for (SmallVector<AccessorType>::iterator i = candArgs.begin();
+       i != candArgs.end(); ++i)
+    for (SmallVector<AccessorType>::iterator j = i + 1; j != candArgs.end();
+         ++j)
+      if (isCandidateAccessorPair(*i, *j))
+        accessorPairs.push_back({*i, *j});
+}
+
+void FunctionSpecializationPass::versionCall(CallOpInterface call) const {
+  SmallVector<AccessorPairType> accessorPairs;
+  getRequireNoOverlapAccessorPairs(call, accessorPairs);
+  if (accessorPairs.empty())
+    return;
+  OpBuilder builder(call);
+  std::unique_ptr<VersionCondition> condition =
+      VersionConditionBuilder(accessorPairs, builder, call->getLoc())
+          .createCondition();
+  CallVersionBuilder(call).version(*condition);
+}
+
+void FunctionSpecializationPass::runOnOperation() {
+  SmallVector<FunctionOpInterface> candidates;
+  getOperation()->walk([&](FunctionOpInterface func) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Processing function \"" << func.getName() << "\"\n");
+    if (isCandidateFunction(func))
+      candidates.push_back(func);
+  });
+
+  ModuleOp module = getOperation();
+  SymbolTableCollection symTable;
+  SymbolUserMap userMap(symTable, module);
+
+  for (FunctionOpInterface func : candidates) {
+    FunctionOpInterface clonedFunc = cloneFunction(func);
+    specializeFunction(clonedFunc);
+
+    for (Operation *op : userMap.getUsers(func)) {
```

Comment:
What happens if you have 2 call sites (one from a kernel and one from some other functions) and you want to version the former but not the latter ?  

I think the pass should have a way to collect candidate call sites and their associated versioning condition. Then go and version only the candidate call site it collected.